### ISSUE TITLE
Memory needed during file upload

### ISSUE THREAD

**Issue by harryve** [pistacheio/pistache#798]
Hi,

I have created a REST server for uploading files. The code is similar to the code by xoac in issue #105. My problem is that the pistache library allocates a lot of memory during the upload. After the file is written to disk, the memory is freed. 
The allocated amount of memory is about ten times the size of the file. This means that I cannot upload files of 50 MB on my embedded system because I run out of memory.
Is there any way to fix this problem?

**Comment by dennisjenkins75** [comment]
Please show us the code where you initialize the Endpoint, and what options you pass in.

**Comment by harryve** [comment]
Here is my code:
```
class UpdateServer {
public:
    explicit UpdateServer(Address addr)
        : httpEndpoint(std::make_shared<Http::Endpoint>(addr))
    {
    }

    void init(size_t maxRequestSize = 4096)
    {
        auto opts = Http::Endpoint::options().maxRequestSize(maxRequestSize);
        httpEndpoint->init(opts);
        setupRoutes();
    }

    void start()
    {
        httpEndpoint->setHandler(router.handler());
        httpEndpoint->serve();
    }

private:
    void setupRoutes()
    {
        using namespace Rest;

        Routes::Post(router, "/ready", Routes::bind(&Generic::handleReady));
        Routes::Post(router, "/file", Routes::bind(&Generic::handleFile));
    }

    std::shared_ptr<Http::Endpoint> httpEndpoint;
    Rest::Router router;
};

int main(int argc, char *argv[])
{
    Port port(9080);
    Address addr(Ipv4::any(), port);

    cout << "Cores = " << hardware_concurrency() << endl;
    UpdateServer updateServer(addr);
    updateServer.init(50*1024*1024);
    updateServer.start();
}
```

**Comment by hyperxor** [comment]
@harryve Hello, can you show `handleReady` and `handleFile` handlers implementation?

**Comment by harryve** [comment]
Of course (this is actually the first part of my code):
```#include <pistache/http.h>
#include <pistache/router.h>
#include <pistache/endpoint.h>
#include <fstream>
#include <chrono>
#include <iostream>

using namespace Pistache;
using namespace std;

namespace Generic {
void handleReady(const Rest::Request&, Http::ResponseWriter response)
{
    cout << "Ready" << endl;
    response.send(Http::Code::Ok); //, "1");
}

void writeToFile(const string& filename, const string& data)
{
    ofstream out;
    out.open(filename.c_str());
    out << data;
    out.close();
}

void handleFile(const Rest::Request& req, Http::ResponseWriter response)
{
    cout << "File" << endl;

    auto start = std::chrono::steady_clock::now();
    writeToFile("test.bin", req.body());
    auto delta = std::chrono::steady_clock::now() - start;
    cout << "File written in: " << std::chrono::duration_cast<std::chrono::milliseconds>(delta).count() << " ms" << endl;
    response.send(Http::Code::Ok);
}
}
```

**Comment by dennisjenkins75** [comment]
Its been so long since I've looked at this code.  At first I thought that maybe pistache was doing some dumb buffer copying to grow the buffer.  But inside "src/common/http.cc", "BodyStep::parseContentLength", we see that once it has the content length, it calls "message->body_.reserve(contentLength);"

Maybe that code is broken.

Please try this: Run your server under valgrind and check for memory leaks.  I suspect that you won't find any major ones. (I test my own app under valgrind and don't have leaks that grow; just some small one-time init leakage).

I wonder if you could link against a different malloc() implementation; one that might be more aggressive about returning freed heap memory back to the OS.

I assume that if you run your server under "strace" (strace -o /tmp/foo.log -xx my-server), and examine the logs, you'll see a fair bit of mmap() and munmap(), but you'll observer that freed heap ram is not returned to the OS.  A 50MB buffer should definitely come directly from mmap() and not sbrk() style heaps.  The last time that i checked (long time ago), the threshold from switching from sbrk() to mmap() was 128kB.

**Comment by harryve** [comment]
The problem that I see is not a memory leak, but a huge amount of memory that is used during the file upload.

I ran my test program with valgrind. If I upload a file of 1k, I see this:
```==16676==     in use at exit: 10,192 bytes in 75 blocks
==16676==   total heap usage: 343 allocs, 268 frees, 127,046 bytes allocated
```
If I upload a file of 10M, I see:
```==16694==     in use at exit: 10,192 bytes in 75 blocks
==16694==   total heap usage: 5,870 allocs, 5,795 frees, 124,077,050 bytes allocated
```

I have also tried strace, but there is no difference in the output (except the virtual addresses of course) when uploading a small or a large file. This is not what I expected. 
I will have a closer look in a couple of days and report my findings.
